spec/helpers/http_mock/debug_port.lua

```diff
@@ -0,0 +1,100 @@
+local helpers = require "spec.helpers"
+local http = require "resty.http"
+local cjson = require "cjson"
+local match = string.match
+local ipairs = ipairs
+local insert = table.insert
+local assert = assert
+
+local http_mock = {}
+
+-- POST as it's not idempotent
+local get_record_param = {
+  method = "POST",
+  path = "/records",
+  headers = {
+    ["Host"] = "mock_debug"
+  }
+}
+
+local get_status_param = {
+  method = "GET",
+  path = "/status",
+  headers = {
+    ["Host"] = "mock_debug"
+  }
+}
+
+function http_mock:setup_debug(debug_param)
+  local debug_port = helpers.get_available_port()
+  local debug_client = http.new()
+  local debug_connect = {
+    scheme = "http",
+    host = "localhost",
+    port = debug_port,
+  }
+
+  self.debug = {
+    port = debug_port,
+    client = debug_client,
+    connect = debug_connect,
+    param = debug_param,
+  }
+end
+
+function http_mock:debug_connect()
+  local debug = self.debug
+  local client = debug.client
+  assert(client:connect(debug.connect))
+  return client
+end
+
+function http_mock:get_record_json()
+  local debug = self:debug_connect()
+  local res = assert(debug:request(get_record_param))
+  assert(res.status == 200)
+  local body = assert(res:read_body())
+  debug:close()
+  return body
+end
+
+function http_mock:get_record()
+  local new_records = cjson.decode(self:get_record_json())
+  for _, record in ipairs(new_records) do
+    insert(self.records, record)
+  end
+
+  return new_records
+end
+
+function http_mock:wait_until_no_request(timeout)
+  local debug = self:debug_connect()
+
+  -- wait until we have no requests on going
+  helpers.wait_until(function()
+    local res = assert(debug:request(get_status_param))
+    assert(res.status == 200)
+    local body = assert(res:read_body())
+    local reading, writing, _ = match(body, "Reading: (%d+) Writing: (%d+) Waiting: (%d+)")
+    -- the status is the only request
+    return assert(reading) + assert(writing) <= 1
+  end, timeout)
+end
+
+function http_mock:get_all_records(timeout)
+  self:wait_until_no_request(timeout)
+  self:get_record()
```

Comment:
`get_record` sounds as if it would return something, but it seems to have important side effects. `retrieve_request_logs` or something in that direction would make it clearer.